### DISCUSSION TITLE
Convert an image from http with 301 Moved Permanently

### DISCUSSION THREAD

**Discussion by Piorko** [ImageMagick/ImageMagick#6721] [Contribute]
Hi,
I'm not able to convert the image via the http protocol:
`error : Unknown IO error
convert-im6.q16: no decode delegate for this image format '' @ error/constitute.c/ReadImage/504.
convert-im6.q16: no data returned 'http://aa.bb/xx/xxx.jpg' @ error/url.c/ReadURLImage/247.
convert-im6.q16: no images defined '/home/xxx/test.jpg' @ error/convert.c/ConvertImageCommand/3258.
` 

When I send a request using curl, I receive the status code 301, with html body:
`> GET /xx/xxx.jpg HTTP/1.1
> Host: aa.bb
> User-Agent: curl/7.52.1
> Accept: */*
>
< HTTP/1.1 301 Moved Permanently
< Server: nginx
< Date: Sat, 21 Dec 2019 23:59:33 GMT
< Content-Type: text/html
< Content-Length: 178
< Connection: keep-alive
< Location: https://aa.bb/xx/xxx.jpg
<
<html>
<head><title>301 Moved Permanently</title></head>
<body bgcolor="white">
<center><h1>301 Moved Permanently</h1></center>
<hr><center>nginx</center>
</body>
</html>
`

Of course adding the -L parameter helps and I receive body of jpg file.

Is it possible to convert an image without manually changing the protocol to https?

ImageMagick version: ImageMagick 6.9.7-4 Q16 x86_64
Environment: Debian 9
Regards

**Comment by fmw42** [comment]
What was your exact command showing the HTTP file path? The error message said the file was moved to a different location. So it is not available at the http path you specified.  Does your Imagemagick policy.xml file permit you to use HTTP?

**Comment by Piorko** [comment]
My Imagemagick policy.xml file permits to use http. I am able to process others http paths.
The problem occurs when I get the answer with 301 status code.

`curl -v 'http://aa.bb/xx/xxx.jpg'` prints html body with 301 Moved Permanently in the header
`curl -v -L 'http://aa.bb/xx/xxx.jpg'` prints the jpeg body

`convert "http://aa.bb/xx/xxx.jpg" "/home/xxx/test.jpg"` doesn't work with the error described above
`convert "https://aa.bb/xx/xxx.jpg" "/home/xxx/test.jpg"` works and converts an image

It looks like IM cant fallow redirection. Example url: http://letsencrypt.org/images/3.jpg

**Comment by fmw42** [comment]
Yes, I think that is true. I cannot deal with redirection. But I will leave it to the developers to comment further or correct me.

**Comment by davemetzak** [comment]
I'm also having the same problem with http not redirecting. Is there a way to disable the built in http handler and instead set it up as a delegate in a similar fashion to how https is used?
